### PR TITLE
Add FFI methods for creating a transaction via PCZT proving & signing.

### PR DESCRIPTION
**Pull request by nuttycom**

### CODE REVIEW

**Review by daira — DISMISSED** [review]
utACK

File: rust/src/lib.rs [hunk 1]
```diff
@@ -2932,6 +2932,252 @@ pub unsafe extern "C" fn zcashlc_create_proposed_transactions(
...
+                account_id,
+                OvkPolicy::Sender,
+                &proposal,
+            )
+            .map_err(|e| anyhow!("Error creating PCZT from single-step proposal: {}", e))?;
+
+            Ok(FfiBoxedSlice::some(pczt.serialize()))
+        } else {
+            Err(anyhow!(
+                "Multi-step proposals are not yet supported for PCZT generation."
```
Comment (daira) [inline rust/src/lib.rs:3009]: Should we actually be returning an array of PCZTs?

File: rust/src/lib.rs [hunk 2]
```diff
@@ -2932,6 +2932,252 @@ pub unsafe extern "C" fn zcashlc_create_proposed_transactions(
...
+///
+/// - `db_data` must be non-null and valid for reads for `db_data_len` bytes, and it must
+///   have an alignment of `1`. Its contents must be a string representing a valid system
+///   path in the operating system's preferred representation.
+/// - The memory referenced by `db_data` must not be mutated for the duration of the
+///   function call.
+/// - The total size `db_data_len` must be no larger than `isize::MAX`. See the safety
+///   documentation of `pointer::offset`.
+/// - `pczt_with_proofs_ptr` must be non-null and valid for reads for `pczt_with_proofs_len` bytes, and it
+///   must have an alignment of `1`. Its contents must be an encoded Proposal protobuf.
```
Comment (daira) [inline rust/src/lib.rs:3092]: > Its contents must be an encoded Proposal protobuf.

I don't think that's a safety requirement?

File: rust/src/lib.rs [hunk 3]
```diff
@@ -2932,6 +2932,252 @@ pub unsafe extern "C" fn zcashlc_create_proposed_transactions(
...
+/// - The memory referenced by `pczt_with_proofs_ptr` must not be mutated for the duration of the
+///   function call.
+/// - The total size `pczt_with_proofs_len` must be no larger than `isize::MAX`. See the safety
+///   documentation of `pointer::offset`.
+/// - `pczt_with_sigs_ptr` must be non-null and valid for reads for `pczt_with_sigs_len` bytes, and it
+///   must have an alignment of `1`. Its contents must be an encoded Proposal protobuf.
```
Comment (daira) [inline rust/src/lib.rs:3098]: Ditto.

File: rust/src/lib.rs [hunk 4]
```diff
@@ -2932,6 +2932,252 @@ pub unsafe extern "C" fn zcashlc_create_proposed_transactions(
...
+///   function call.
+/// - The total size `db_data_len` must be no larger than `isize::MAX`. See the safety
+///   documentation of `pointer::offset`.
+/// - `pczt_with_proofs_ptr` must be non-null and valid for reads for `pczt_with_proofs_len` bytes, and it
+///   must have an alignment of `1`. Its contents must be an encoded Proposal protobuf.
+/// - The memory referenced by `pczt_with_proofs_ptr` must not be mutated for the duration of the
+///   function call.
+/// - The total size `pczt_with_proofs_len` must be no larger than `isize::MAX`. See the safety
+///   documentation of `pointer::offset`.
+/// - `pczt_with_sigs_ptr` must be non-null and valid for reads for `pczt_with_sigs_len` bytes, and it
```
Comment (daira) [inline rust/src/lib.rs:3097]:
```suggestion
/// - `pczt_with_proofs_ptr` must be non-null and valid for reads for `pczt_with_proofs_len` bytes, and it
```

File: rust/src/lib.rs [hunk 5]
```diff
@@ -2932,6 +2932,252 @@ pub unsafe extern "C" fn zcashlc_create_proposed_transactions(
...
+/// - The memory referenced by `pczt_with_sigs_ptr` must not be mutated for the duration of the
+///   function call.
+/// - The total size `pczt_with_sigs_len` must be no larger than `isize::MAX`. See the safety
+///   documentation of `pointer::offset`.
+/// - `pczt_with_sigs_ptr` must be non-null and valid for reads for `pczt_with_sigs_len` bytes, and it
+///   must have an alignment of `1`. Its contents must be an encoded Proposal protobuf.
+/// - The memory referenced by `pczt_with_sigs_ptr` must not be mutated for the duration of the
+///   function call.
+/// - The total size `pczt_with_sigs_len` must be no larger than `isize::MAX`. See the safety
+///   documentation of `pointer::offset`.
```
Comment (daira) [inline rust/src/lib.rs:3108]: These lines seem to be a duplicate of the ones above.
```suggestion
```

File: rust/src/lib.rs (hunk 1 above)
Comment (nuttycom) [reply rust/src/lib.rs:3009]: For a multi-step proposal, we will likely want a different API; there will be some cases where we can construct all of the PCZTs for the steps at once, but that's not the case for every multi-step proposal (for example, if the first step requires a multisig PCZT, I don't think we can necessarily construct the PCZT for the second step.)

File: rust/src/lib.rs [hunk 6]
```diff
@@ -2932,6 +2932,252 @@ pub unsafe extern "C" fn zcashlc_create_proposed_transactions(
...
+/// - `spend_params` must be non-null and valid for reads for `spend_params_len` bytes,
+///   and it must have an alignment of `1`. Its contents must be the Sapling spend proving
+///   parameters.
```
Comment (daira) [inline rust/src/lib.rs:3111]: > Its contents must be the Sapling spend proving parameters.

I don't think this is a safety requirement.

File: rust/src/lib.rs [hunk 7]
```diff
@@ -2932,6 +2932,252 @@ pub unsafe extern "C" fn zcashlc_create_proposed_transactions(
...
+/// - The memory referenced by `spend_params` must not be mutated for the duration of the
+///   function call.
+/// - The total size `spend_params_len` must be no larger than `isize::MAX`. See the safety
+///   documentation of `pointer::offset`.
+/// - `output_params` must be non-null and valid for reads for `output_params_len` bytes,
+///   and it must have an alignment of `1`. Its contents must be the Sapling output
+///   proving parameters.
```
Comment (daira) [inline rust/src/lib.rs:3118]: > Its contents must be the Sapling output proving parameters.

I don't think this is a safety requirement.

File: rust/src/lib.rs (hunk 6 above)
Comment (nuttycom) [reply rust/src/lib.rs:3111]: It's also wrong; I copied this documentation from elsewhere but this is a path. I was actually wondering whether this should be passed as a CStr instead.

**Review by daira — APPROVED** [review]
utACK with comments. The comments on safety documentation are blocking (if I'm correct).

File: rust/src/lib.rs (hunk 6 above)
Comment (daira) [reply rust/src/lib.rs:3111]: We have other places where we pass a path to parameter files and it should match those, I think.

Comment (nuttycom) [reply rust/src/lib.rs:3111]: I'm not sure why those other paths aren't passed as C strings.

File: rust/src/lib.rs (hunk 2 above)
Comment (nuttycom) [reply rust/src/lib.rs:3092]: Note that although it's not a safety requirement, we include what the contents are supposed to be in the safety documentation everywhere (see also the path to `db_data` etc.). Duplicating the parameter documentation seems excessive to me.

File: rust/src/lib.rs (hunk 6 above)
Comment (daira) [reply rust/src/lib.rs:3111]: Actually, loading in the parameters from a file is quite expensive iirc. Don't we have a way to preload them?

**Review by daira — APPROVED** [review]
utACK

**Pull request merged**